shape.h

```diff
@@ -33,21 +33,24 @@ typedef uint32_t redblack_id_t;
 # define SHAPE_MAX_VARIATIONS 8
 
 # define INVALID_SHAPE_ID SHAPE_MASK
-# define ROOT_SHAPE_ID 0x0
 
-# define SPECIAL_CONST_SHAPE_ID (ROOT_SHAPE_ID + 1)
-# define OBJ_TOO_COMPLEX_SHAPE_ID (SPECIAL_CONST_SHAPE_ID + 1)
-# define FIRST_T_OBJECT_SHAPE_ID (OBJ_TOO_COMPLEX_SHAPE_ID + 1)
+#define ROOT_SHAPE_ID               0x0
+#define SPECIAL_CONST_SHAPE_ID      0x1
+//      ROOT_TOO_COMPLEX_SHAPE_ID   0x2
+#define FIRST_T_OBJECT_SHAPE_ID     0x3
+
+extern ID ruby_internal_object_id;
 
 typedef struct redblack_node redblack_node_t;
 
 struct rb_shape {
     struct rb_id_table *edges; // id_table from ID (ivar) to next shape
     ID edge_name; // ID (ivar) for transition from parent to rb_shape
-    attr_index_t next_iv_index;
+    attr_index_t next_field_index; // Fields are either ivars or internal properties like `object_id`
     attr_index_t capacity; // Total capacity of the object with this shape
     uint8_t type;
     uint8_t heap_index;
+    uint8_t flags;
     shape_id_t parent_id;
     redblack_node_t *ancestor_index;
 };
```

Comment:
What is the size of rb_shape now?  We might want to move `flags` to the end of the struct because I think this will leave wasted space between `flags` and `parent_id`.

On master it's 40 bytes:

```
(lldb) p sizeof(struct rb_shape)
(unsigned long) 40
```

I'm not sure how much we care about shape memory usage, but maybe we could split the type field. We might only need 4 bits for `type`.

Reply:
It's unchanged, I've put the flags here because there was some padding:

```diff
diff --git a/shape.c b/shape.c
index fb739d3614..8aec985496 100644
--- a/shape.c
+++ b/shape.c
@@ -1434,6 +1434,7 @@ Init_default_shapes(void)
 void
 Init_shape(void)
 {
+    fprintf(stderr, "sizeof(rb_shape_t) = %lu\n", sizeof(rb_shape_t));
 #if SHAPE_DEBUG
     /* Document-class: RubyVM::Shape
      * :nodoc: */
```

```
$ make -j run
sizeof(rb_shape_t) = 40
```

> I'm not sure how much we care about shape memory usage

Not that much, but I think we could save some spaces by not using a full pointer for `redblack_node_t *ancestor_index;`. We could just have an offset. I think It would be nice if shapes were 32B instead of 40B.